components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,366 @@
+    ///
+    /// Returns an error if the serialized string fails to parse.
+    /// The serialized string requires starting capacity integer, followed by space delimited
+    /// integer code points. There must be an even number of elements (not including the
+    /// capacity int), and must be in ascending sorted order.
+    ///
+    /// Example String: `"4 0 5 10 15"` designates a capacity of size `4`, followed by 2 ranges
+    /// The ranges are `{0, 4}` and `{10, 14}` inclusive
+    pub fn new(serialize: &str) -> Result<UnicodeSet, Box<dyn Error>> {
+        match parse_serial_string(serialize) {
+            Ok(serialize) => Ok(UnicodeSet {
+                inv_list: serialize,
+            }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Returns Result of UnicodeSet from a single pair of integers defining a range
+    ///
+    /// `start`: inclusive, `end`: exclusive
+    ///
+    /// Returns an error if the range is invalid (out of order and out of bounds).
+    ///
+    /// Example Call: `UnicodeSet::from_range(&0, &15)`
+    pub fn from_range(start: u32, end: u32) -> Result<UnicodeSet, Box<dyn Error>> {
+        if start > end {
+            return Err(format!("Range is out of order. start: {} end: {}", start, end).into());
+        }
+        if start < CODEPOINT_MIN || end > CODEPOINT_MAX {
+            return Err(format!(
+                "Range is out of bounds. start: {}, min: {}, end: {}, max: {}",
+                start, CODEPOINT_MIN, end, CODEPOINT_MAX
+            )
+            .into());
+        }
+        Ok(UnicodeSet {
+            inv_list: vec![start, end],
+        })
+    }
+
+    /// Returns UnicodeSet spanning entire Unicode range
+    ///
+    /// The range spans from `0x0 -> 0x10FFFF` inclusive
+    pub fn all() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, CODEPOINT_MAX + 1],
+        }
+    }
+
+    /// Returns UnicodeSet spanning BMP range
+    ///
+    /// The range spans from `0x0 -> 0xFFFF` inclusive
+    pub fn bmp() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, BMP_MAX + 1],
+        }
+    }
+
+    /// Yields an iterator of start and stop points of ranges in the UnicodeSet
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu4x_unicodeset::UnicodeSet;
+    /// let example = UnicodeSet::new("4 0 10 15 20");
+    /// let mut example_iter = example.iter();
+    /// example_iter.next(); // => 0
+    /// example_iter.next(); // => 10
+    /// example_iter.next(); // => 10, etc.
+    /// ```
+    pub fn iter(&self) -> Iter<u32> {
+        self.inv_list.iter()
+    }
+
+    /// Returns the number of elements of the UnicodeSet
+    ///
+    /// Returns an error if a complete range is not defined in the UnicodeSet
+    pub fn size(&self) -> usize {
+        if self.is_empty() {
+            return 0;
+        }
+        let end_point_sum: u32 = self.iter().skip(1).step_by(2).sum::<u32>();
```

Comment:
The worst case sum is when the inversion list contains every integer between 0 and char::MAX.  In that case, the sum will be something like (char::MAX^2)/4, which is approximately 3.1e11.  However, u32::MAX is 4.3e9, which is not big enough.

Actually, I would like to see you write this such that you only need to iterate over the vector one time, instead of two times.  You can look at two adjacent entries at the same time, and either add or subtract the difference.

Reply:
The idea of iterating over the single vector was my original idea but it seemed to keep them in iterators. But the first point you raised is right and so agreed and changed.